File: src/tvision4c/KeyDownEvent.cpp
#include "KeyDownEvent.h"
#include "Error.h"
#include "Point.h"
#include <cstring>

EXPORT tv::Error TV_KeyDownEvent_new(KeyDownEvent** out) {
    return tv::checkedNew(out);
}

EXPORT tv::Error TV_KeyDownEvent_delete(KeyDownEvent* self) {
    return tv::checkedDelete(self);
}

EXPORT tv::Error TV_KeyDownEvent_equals(KeyDownEvent* self, KeyDownEvent* other, BOOL* out) {
    if (!out) {
        return tv::Error_ArgumentNull;
    }

    if (!self && !other) {
        *out = TRUE;
        return tv::Success;
    }

    if (!self || !other) {
        *out = FALSE;
        return tv::Success;
    }

    if (self->keyCode != other->keyCode || self->controlKeyState != other->controlKeyState ||
        self->textLength != other->textLength) {
        *out = FALSE;
        return tv::Success;
    }

    // Compare text content up to textLength
    if (self->textLength > 0) {
        *out = std::memcmp(self->text, other->text, self->textLength) == 0;
        return tv::Success;
    }

    *out = TRUE;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_hash(KeyDownEvent* self, int32_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    // Hash the keyCode, controlKeyState, and textLength
    int32_t result = 0;
    tv::hash(self->keyCode, &result);
    tv::hash(self->controlKeyState, &result);
    tv::hash(self->textLength, &result);

    // Include text content in hash if present
    if (self->textLength > 0) {
        for (uint8_t i = 0; i < self->textLength; i++) {
            tv::hash((uint8_t)self->text[i], &result);
        }
    }

    *out = result;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_get_keyCode(KeyDownEvent* self, uint16_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->keyCode;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_set_keyCode(KeyDownEvent* self, uint16_t value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->keyCode = value;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_get_charCode(KeyDownEvent* self, uint8_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->charScan.charCode;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_set_charCode(KeyDownEvent* self, uint8_t value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->charScan.charCode = value;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_get_scanCode(KeyDownEvent* self, uint8_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->charScan.scanCode;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_set_scanCode(KeyDownEvent* self, uint8_t value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->charScan.scanCode = value;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_get_controlKeyState(KeyDownEvent* self, uint16_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->controlKeyState;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_set_controlKeyState(KeyDownEvent* self, uint16_t value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->controlKeyState = value;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_get_text(KeyDownEvent* self, char** out, uint8_t* outTextLength) {
    if (!self || !out || !outTextLength) {
        return tv::Error_ArgumentNull;
    }

    *outTextLength = self->textLength;
    *out = self->text;
    return tv::Success;
}

EXPORT tv::Error TV_KeyDownEvent_set_text(KeyDownEvent* self, char* value, uint8_t textLength) {
    if (!self || (!value && textLength > 0)) {
        return tv::Error_ArgumentNull;
    }

    if (textLength > maxCharSize) {
        return tv::Error_BufferTooSmall;
    }

    self->textLength = textLength;
    memset(self->text, 0, maxCharSize);

    if (textLength > 0) {
        memcpy(self->text, value, textLength);
    }

    return tv::Success;
}

File: src/TurboVision/System/KeyDownEvent.cs
using System.Runtime.InteropServices;
using System.Text;

namespace TurboVision.System;

public unsafe partial class KeyDownEvent(void* ptr, bool owned, bool placement)
    : NativeObject<KeyDownEvent>(ptr, owned, placement)
{
    private sealed class Factory : NativeObjectFactory<Factory>
    {
        public Factory()
            : base(
                NativeMethods.TV_KeyDownEvent_placementSize,
                NativeMethods.TV_KeyDownEvent_placementNew,
                NativeMethods.TV_KeyDownEvent_new
            ) { }
    }

    public static int PlacementSize => Factory.Instance.PlacementSize;

    public KeyDownEvent(byte* placement)
        : this(Factory.Instance.PlacementNew(placement), owned: true, placement: true) { }

    public KeyDownEvent()
        : this(Factory.Instance.New(), owned: true, placement: false) { }

    protected override void PlacementDeleteCore(void* ptr)
    {
        TurboVisionException.Check(NativeMethods.TV_KeyDownEvent_placementDelete(ptr));
    }

    protected override void DeleteCore()
    {
        TurboVisionException.Check(NativeMethods.TV_KeyDownEvent_delete(Ptr));
    }

    protected override bool EqualsCore(KeyDownEvent other)
    {
        TurboVisionException.Check(
            NativeMethods.TV_KeyDownEvent_equals(Ptr, other.Ptr, out var equals)
        );
        return equals;
    }

    protected override int GetHashCodeCore()
    {
        TurboVisionException.Check(NativeMethods.TV_KeyDownEvent_hash(Ptr, out var hash));
        return hash;
    }

    public ushort KeyCode
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(
                NativeMethods.TV_KeyDownEvent_get_keyCode(Ptr, out var keyCode)
            );
            return keyCode;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_KeyDownEvent_set_keyCode(Ptr, value));
        }
    }

    public byte CharCode
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(
                NativeMethods.TV_KeyDownEvent_get_charCode(Ptr, out var charCode)
            );
            return charCode;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_KeyDownEvent_set_charCode(Ptr, value));
        }
    }

    public byte ScanCode
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(
                NativeMethods.TV_KeyDownEvent_get_scanCode(Ptr, out var scanCode)
            );
            return scanCode;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_KeyDownEvent_set_scanCode(Ptr, value));
        }
    }

    public ushort ControlKeyState
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(
                NativeMethods.TV_KeyDownEvent_get_controlKeyState(Ptr, out var controlKeyState)
            );
            return controlKeyState;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(
                NativeMethods.TV_KeyDownEvent_set_controlKeyState(Ptr, value)
            );
        }
    }

    public string Text
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            unsafe
            {
                TurboVisionException.Check(
                    NativeMethods.TV_KeyDownEvent_get_text(Ptr, out var textPtr, out var textLength)
                );
                return Encoding.UTF8.GetString(textPtr, textLength);
            }
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            var textBytes = Global.UTF8Encoding.GetBytes(value);
            if (textBytes.Length > byte.MaxValue)
                throw new ArgumentException("Text is too long.", nameof(value));
            TurboVisionException.Check(
                NativeMethods.TV_KeyDownEvent_set_text(Ptr, textBytes, (byte)textBytes.Length)
            );
        }
    }

    internal static partial class NativeMethods
    {
        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_placementSize(
            out int outSize,
            out int outAlignment
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_placementNew(byte* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_placementDelete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_new(out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_delete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_equals(
            void* self,
            void* other,
            [MarshalAs(UnmanagedType.I4)] out bool @out
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_hash(void* self, out int @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_get_keyCode(void* self, out ushort @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_set_keyCode(void* self, ushort value);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_get_charCode(void* self, out byte @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_set_charCode(void* self, byte value);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_get_scanCode(void* self, out byte @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_set_scanCode(void* self, byte value);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_get_controlKeyState(
            void* self,
            out ushort @out
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_set_controlKeyState(void* self, ushort value);

        [LibraryImport(Global.DLL_NAME)]
        public static unsafe partial Error TV_KeyDownEvent_get_text(
            void* self,
            out byte* @out,
            out byte outTextLength
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_KeyDownEvent_set_text(
            void* self,
            Span<byte> value,
            byte textLength
        );
    }
}
